### ISSUE TITLE
Strong web authentication / 2FA authentication

### ISSUE THREAD

**Issue by yunohost-bot** [YunoHost/issues#238]
###### Original Redmine Issue: [237](https://dev.yunohost.org/issues/237)

Author Name: **moul**

---

Strong web authentication could be integrated on admin and users portals. To do so client certificate could be used.

We could let the choice to the admin or to users to use password or certificate authentication.

The inconvenient is that we can't connect to YunoHost web interface from a machine which don't have this certificate.

[Discussion about password and certificate authentication](https://security.stackexchange.com/questions/3605/certificate-based-authentication-vs-username-and-password-authentication).

**Comment by randomstuff** [comment]
Maybe [WebAuth](https://www.w3.org/TR/webauthn/) support?

**Comment by Psycojoker** [comment]
I don't know WebAuth but I know that double factor kind of auth is also popular, I'm not sure on how close to that it is.

**Comment by randomstuff** [comment]
WebAuthn is supposed to be a (very new) standard for token based authentication: using something like a authentication USB/BlueTooth device to handle stronger/two-factor authentication AFAIU. I didn't get time to understand how that's supposed to work yet and I did not find a easily explanation yet. Support had landed in Firefox and Chromium.

**Comment by Psycojoker** [comment]
Oh, that's for yuby keys and all those things?

**Comment by randomstuff** [comment]
Yes, for example. You're supposed to be able to use a smartphone instead. Apparently there's supposed to be a software token implemented in Firefox (webauthn_enable_softtoken). I didn't manage to get it work yet though.

**Comment by Psycojoker** [comment]
That could indeed be very interesting. But I'm not really expecting people to start working on it in the core team because I don't think anyone uses that kind of tool :/

**Comment by Torenga** [comment]
The lack of 2fa is a real bummer. While Nextcloud support it from within the script, the SSO renders it moot unless one creates individual users on i.e. Nextcloud, rendering the SSO moot! And there's still (iirc) a hardcoded, no-username password prompt for logging into the web-ui. :O

Unfortunately I want the extra security 2fa provides more than all the other Yunohost apps combined - so I'm staying on NextcloudPi for now. Hope to see this as a feature in the future.

**Comment by alexAubin** [comment]
We were quickly discussing this the other day with somebody else ... Could you provide more info on what kind of channel(s) you'd imagine for the 2FA mechanism ? (For example, depending on a Google-based 2FA mechanism would be kind of a big no-no...

**Comment by randomstuff** [comment]
What about TOTP, [RFC6238](https://tools.ietf.org/html/rfc6238)?

**Comment by Torenga** [comment]
[TOTP](https://en.wikipedia.org/wiki/Time-based_One-time_Password_algorithm) based on [RFC 6238](https://tools.ietf.org/html/rfc6238) is one route.

**Comment by alexAubin** [comment]
2FA using TOTP sounds indeed legit to me and somewhat ~simple and with little requirements

Only annoying thing is that 90% of tutorials on the internet refer to a google_auth lib (for SSH auth) which is .. ugh

But apparently there may be way to ways to do this without that lib

**Comment by Torenga** [comment]
Perhaps https://github.com/pyauth/pyotp could be used instead of google-auth?

**Comment by randomstuff** [comment]
> Only annoying thing is that 90% of tutorials on the internet refer to a google_auth lib (for SSH auth) which is .. ugh

I thought the case discussed here was strong authentication for the web access, not for SSH.

Anyway, for PAM it seems to be possible to use either:

* libpam-oath
* libpam-google-authenticator

I'm don't believe libpam-google-authenticator is particularly bad. Yes, it's made by Google but I don't believe it means it ships some Google backdoor anyhow.

**Comment by Aeris1One** [comment]
Webauthn is a simple ways to make all FIDO keys (Yubikeys, Solo,... Even smartphones if computer support Bluetooth/NFC) working, at least for a website it's a matter of a few JS code, idk how it would render with LDAP though :/
I think FIDO keys (passwordless auth) and 2FA (like discussed here) are two different things, should I open another issue especially for Webauthn ?

**Comment by artybdrlt** [comment]
Hello,
Do you have any update on this issue? Me and my colleagues from Lydra are interested in this topic as enabling 2FA auth for web panel would ensure a better security level.

**Comment by alexAubin** [comment]
Not much to update ... I think the big picture is : 
- there are some slow progress made to try to get rid of the infamous "admin" special-user and instead have an "admins" group composed of regular yunohost user that would have access to the webadmin ... but still a lot of work to adapt some pieces of code (but imho is a prerequisites to start working on 2FA)
- ... and then we could add some sort of OTP or time-based 2FA thingies tokens to yunohost users, which could then be used when logging in the webadmin ...

**Comment by mateMathieu** [comment]
I can't put more agreement on these two absolutely essential steps that would greatly improve the security of the platform.
Thanks for your great work, you have all my good wish in this effort.
Let us know how we can help.

**Comment by artybdrlt** [comment]
Thanks for the update @alexAubin. 
Looking forward to hearing more about that.

**Comment by ericgaspar** [comment]
I don't know if this implementation is of any use : https://github.com/pyauth/pywarp

**Comment by Bugsbane** [comment]
Pywarp seems to require hardware tokens, which are great for some people, but not for many others. Depending on users buying a hardware token will just mean many users skip 2FA, weaking Yunohost's security. While a 2FA solution is definitely needed (after the rework Alex mentioned), to be useful to the largest number of users, it should support some form of software 2Fa (eg Authy / Google Authenticator etc). Being able to use hardware tokens as well would just be a nice bonus.

**Comment by Aeris1One** [comment]
WebAuthn is all about hardware token, but by hardware tokens it also includes Android phones with TPM/Google Titan chip (most of new phones) as well as all iThings with TouchID/FaceID and Windows 11 computers with TPM. There's also Yubikeys/Hardware authenticators but they are minoritarian as of now. Goal of W3C is to slowly switch completely from passwords to passwordless authentication using more reliable hardware systems.
Yunohost is still one of the only open source self-hosting software that does not implement at least software authentication, I know it's a volunteer project and that revamping the whole authentication system is not the funniest code thing to do but i think strong authentication should be in the priority list.

**Comment by Bugsbane** [comment]
Agreed. Most people using Yunohost do so for privacy reasons, I would be, but as the old saying goes, there's no privacy without security. Not having 2FA to access all of ones self hosted software and data, is a pretty serious security minimization.

I would gladly contribute towards a bounty on this.

**Comment by WoozyFace** [comment]
I also opened a topic on this one! It's not a thing we want, but it's a thing we NEED! Like coffee, a croissant, or a pain in the morning!

**Comment by ticktostick** [comment]
its very welcome
Keycloak+privacyIDEA

**Comment by jbastruz** [comment]
Hi everyone, any news about this feature, i'm quite interesting to upgrade my server's security :) ?

**Comment by Aeris1One** [comment]
May I suggest setting this to "Priority: High" or at least Medium ?
Yunohost is all about privacy, and privacy doesn't exist without security.

It's way easier to set-up a brute-force on every .ynh.fr/.nohost.me/... domains than to grab access to a Google account because it's now an obligation to setup 2FA on Google accounts !

I'm also ready, as @Bugsbane said, to contribute to a bounty on this issue.

**Comment by remmah** [comment]
On a related note, passwordless authentication through WebAuthn is about to pick up steam through Apple's [Passkeys](https://developers.apple.com/videos/play/wwdc2022/10092/) implementation (the linked video has information on what needs to be supported server-side). 1Password also recently joined the FIDO Alliance and teased passwordless login via WebAuthn.

At this point, it may be worth leapfrogging 2FA and focusing on passwordless WebAuthn support.

**Comment by BRGustavoRibeiro** [comment]
On my opinion, this should be **high priority**, if not immediate.
Our servers are extremely insecure without 2FA and reCAPTCHA.

**Comment by hex-m** [comment]
> Keycloak+privacyIDEA

This points to the right direction imho. YunoHost is currently using LDAP for SSO. Modern web authentication today usually means delegating the authentication flow to an IDP (Identity Provider) using SAML or OIDC (OpenID Connect). Keycloak would be one option of an IDP that could be used. related: https://github.com/YunoHost/issues/issues/676

PrivacyIDEA is a "middleware" where users can configure their 2FA-tokens independent of the chosen IDP-Implementation.

**Comment by Cuthbert1337** [comment]
I tried Yunohost and like it really much although nextcloud is already two major versions behind (probably it is hard to keep everything up to date).

But missing 2fa is much worse. I could live with a little outdated software (as long as it is still maintained and secure). But knowing all the admin stuff is only protected by one single password is a no go nowadays.

**Comment by ericgaspar** [comment]
@Cuthbert1337  If you have YunoHost 11 installed, you can get the last Nextcloud 24.0.3 version:
`sudo yunohost app upgrade nextcloud -u https://github.com/YunoHost-Apps/nextcloud_ynh/tree/24.0.0`

That leaves one point left...

**Comment by Bugsbane** [comment]
@ericgaspar Does that give a stable, production ready version of ynh_nextcloud, or is that command providing a testing version, or similar?

**Comment by tituspijean** [comment]
You can follow-up with its development here: https://github.com/YunoHost-Apps/nextcloud_ynh/pull/487. It is still officially testing version, since it is not merged in the main branch.
As of today it passes all automated tests and at least one user confirms it working.
I am tagging our last three messages as off-topic.

**Comment by Mecallie** [comment]
> > Keycloak+privacyIDEA
> 
> This points to the right direction imho. YunoHost is currently using LDAP for SSO. Modern web authentication today usually means delegating the authentication flow to an IDP (Identity Provider) using SAML or OIDC (OpenID Connect). Keycloak would be one option of an IDP that could be used. related: #676
> 
> PrivacyIDEA is a "middleware" where users can configure their 2FA-tokens independent of the chosen IDP-Implementation.

I could not agree more on this. Most applications nowadays can use oauth/saml for authentication too, it would make packaging easier than using ldap directly I imagine?

But other than that: I am really struggling to let users in on my instance knowing that 2FA is not possible with the admin interface and any app that would like to use SSO. I have an own Nextcloud instance, but there is no way I would trust my personal data to anything not featuring 2FA.

Is there anything I can do to help move this along? Small donation or something? I am not a coder, just an admin so not much use there.

**Comment by alexAubin** [comment]
2FA is great and should be enforced everywhere as possible, but I'm always surprised to see that many people being like "I can't live without 2FA" ... I mean a lot of "non-tech-savvy" people are just glad when things works, and the vast majority of them don't use password managers ... (or is this precisely the argument *for* 2FA ?)

Anyway, I guess 2FA for the webadmin is "sort of" on the roadmap (I mean "sort of" considering there's not really any roadmap past "the very-next ynh version")

But 2FA for user portal is still a long way away. And it brings specific challenges : for example Nextcloud instances are recommended to be kept "public" - meaning you don't *have* to log in on the SSO to reach Nextcloud, you can also login through Nextcloud's login page. This is because otherwise this will block desktop and mobile clients I think ?

So I'm curious what's the design to have 2FA for that kind of apps behind a SSO and still be able to use desktop and mobile client ... like, if the desktop / mobile clients can't interface with Yunohost's SSO which handles the 2FA, that seems to defeat the purpose ?

**Comment by Aeris1One** [comment]
> So I'm curious what's the design to have 2FA for that kind of apps behind a SSO and still be able to use desktop and mobile client ... like, if the desktop / mobile clients can't interface with Yunohost's SSO which handles the 2FA, that seems to defeat the purpose ?

@alexAubin 

As an example, I recently set up a LemonLDAP instance to replace Yunohost, auth.mydomain.com. There's a Nextcloud at cloud.mydomain.com.
I'm using SAML to make Nextcloud uses LemonLDAP auth. If you go to cloud.mydomain.com, you'll instantly be redirected to auth.mydomain.com to login, just like how SSOWat works with private apps. Except it's in fact Nextcloud managing this redirection.
If you just use cloud.mydomain.com you'll be redirected and will never see Nextcloud login page, you'll only see LL::NG's one, because Nextcloud will redirect you there. If I send you a public sharing link, though, you'll be able to access the file without Nextcloud redirecting you to the LL:NG login page. This also works with Desktop and Mobile clients.
LL:NG login page supports 2FA. Even if Nextcloud didn't support 2FA, it wouldn't matter, since the login is handled in LL:NG and then forwarded to Nextcloud through SAML.

With SSOWat, if Nextcloud is fully private, you'll only be able to log in through SSOWat, never through Nextcloud login form. But you'll not be able to share links to files or use Desktop/Mobile clients. If Nextcloud is fully public, you'll be able to see the Nextcloud login page and login through it, and you will also be able to use clients and everything else. That second case is not really an SSO, because you'll log in through Nextcloud login page (except if you're using the SSO tile), it's just an LDAP server which is syncing passwords between apps.
SSOWat doesn't support 2FA, and adding 2FA to Nextcloud would break SSO (login using the SSOWat tile)

That would be a HUGE architectural change on how Yunohost is working. But I think it may be a way to go. Not every app supports SAML, but some do support OpenID Connect and/or HTTP Headers, and for apps which support none of them, sticking with LDAP would maybe be possible, but 2FA wouldn't work.

**Comment by alexAubin** [comment]
Eeeh but then can you still configure 2FA on desktop/mobile clients ?

Could you elaborate what piece of configuration you changed in Nextcloud for this ?

**Comment by Aeris1One** [comment]
There's a native SAML integration on Nextcloud, I just had to do some basic setup so it uses my LemonLDAP SAML server. Mostly like the LDAP integration needs to be set up so it uses Yunohost LDAP server.
When login in using Desktop/Mobile, it opens a web browser, redirects me to LL:NG and upon login to it and I being redirected to Nextcloud, Nextcloud asks me “Would you like to allow [device name] to access your Nextcloud accounts?”. I answer yes and go back to the Nextcloud client and, well… I'm logged in!

**Comment by jbastruz** [comment]
sorry guys, I read you use "LemonLDAP" to be able to use 2FA on Yunohost, do you know how to do it ? i didn't find any explanation about that on the internet... ^^'

**Comment by WoozyFace** [comment]
I once did this quest on the forum, Aleks explained there that they will not implement it and that we need to fix something ourselves. #sadface… if someone could make an app for it, then we might have a change. If I could do it I already had.

**Comment by tituspijean** [comment]
Sorry guys, but if you are not contributing to the discussion or the code, I'll mark our messages as off-topic. Regarding "fixing something ourselves", you can have a look to the Dex app. It's not perfect, but it works for Outline for example.

**Comment by WoozyFace** [comment]
What do you mean by outline. With something to fix our selves is that a dev told me on a forum post that if we want 2fa then we should make or script it ourselves. How is this not on topic? Op 5 sep. 2022 om 21:20 heeft tituspijean ***@***.***> het volgende geschreven:﻿
Sorry guys, but if you are not contributing to the discussion or the code, I'll mark our messages as off-topic. Regarding "fixing something ourselves", you can have a look to the Dex app. It's not perfect, but it works for Outline for example.

**Comment by Mecallie** [comment]
@tituspijean can you comment on how much work it is to enable 2FA (preferably via Oauth/SAML or similar so that you can use that for the apps as well)?

The apps are all installed via scripts and use the ldap db for SSO if it exists correct? So why is it so hard to redesign the website? It must be possible to serve a different admin page, leaving the ldap db intact and using the same scripts to install the apps? That way you could migrate slowly to an auth method that supports 2FA for all apps?

Am I missing some big api things that use the webinterface or something?

I am just hoping to get this rolling. I would love to be able to offer Yunohost to businesses, but not having 2FA is killing. From a legal standpoint I would say that I did not use "due diligence" when I offer this to people but cannot provide adequate security ;(

**Comment by ericgaspar** [comment]
Lets not forget that YunoHost project is run by voluntary contributions.

**Comment by WoozyFace** [comment]
> Lets not forget that YunoHost project is run by voluntary contributions.

Never forget this, but even if it is not intended to sell to businesses which is acceptable. Then still it would be a great thing to have even for your own use. Why would you use some kind of 2fa on your bank account, if the developers of that app/site are not using 2fa. It's like giving your credit card and pin code away to hackers.

**Comment by tituspijean** [comment]
@ShaAlexander The Outline app. Both Dex and Outline are in our catalog.
I guess I was a bit harsh on the off-topic-ing.

@Mecallie Sure. Much work. Right now, all requests to apps pass through our home-made SSO. Improve it with 2FA, somehow stored in the LDAP scheme, or replace the SSO, and you are golden.

> It must be possible to serve a different admin page, leaving the ldap db intact and using the same scripts to install the apps? That way you could migrate slowly to an auth method that supports 2FA for all apps?

I have absolutely not understood that. Regarding the webadmin, Aleks has extensively written about it above.

> I am just hoping to get this rolling. I would love to be able to offer Yunohost to businesses, but not having 2FA is killing. From a legal standpoint I would say that I did not use "due diligence" when I offer this to people but cannot provide adequate security ;(

Then help us help you help us all. If you use YunoHost for your business, consider contributing to the code.

**Comment by WoozyFace** [comment]
> @tituspijean can you comment on how much work it is to enable 2FA (preferably via Oauth/SAML or similar so that you can use that for the apps as well)?
> 
> The apps are all installed via scripts and use the ldap db for SSO if it exists correct? So why is it so hard to redesign the website? It must be possible to serve a different admin page, leaving the ldap db intact and using the same scripts to install the apps? That way you could migrate slowly to an auth method that supports 2FA for all apps?
> 
> Am I missing some big api things that use the webinterface or something?
> 
> I am just hoping to get this rolling. I would love to be able to offer Yunohost to businesses, but not having 2FA is killing. From a legal standpoint I would say that I did not use "due diligence" when I offer this to people but cannot provide adequate security ;(

I thought that too. But I guess that you even could use some kind of embedded stuff from Google https://cloud.google.com/identity-platform/docs/web/mfa

**Comment by WoozyFace** [comment]
> @tituspijean  I have absolutely not understood that. Regarding the webadmin, Aleks has extensively written about it above.

https://forum.yunohost.org/t/2fa-for-the-admin-panel-please/18547/10

**Comment by alexAubin** [comment]
>Aleks explained there that they will not implement it and that we need to fix something ourselves

Really ? Would love to know where exactly I said this .. (Who knows, maybe I did ...)

> how much work it is to enable 2FA (preferably via Oauth/SAML [...]) ? [...] why is it so hard to redesign the website?

I mean, I don't even know where to start with this ... "Why is it so hard to completely rework the entire SSO system of YunoHost and its 300 apps ecosystem ?" ... Not to mention that SSOs in general are one hell of a thing to engineer right.

I don't know if you folks realize how depressing this kind of comment is. The other day we had to deal *again* with somebody *toxically* complaining about how user self-registration mechanism is still not there and yet should have been the absolute top number one priority for 9 years (his words). And this was like the 5th time at least he was ranting about it.

Before this, we had another user aggressively complaining about Bullseye support and how this should be the top number one priority.

I mean the good thing is that this shows Yunohost seems to be useful for some folks ? But ultimately these comments don't help with **anything**.

Yunohost is ran by volunteers and we don't really have any super quantitative data on how much work is done, but wouldn't be surprised that this could be valued to at least 100 or 200k€ / year of work. Hell, just the support itself is equivalent to a full-time job. Keeping everything "alive" between app upgrades, infrastructure stuff, upgrading to new Debian versions, fixing bugs, etc : another full-time invisible-ant job. (Not mentioning packaging new apps, and there's basically always 10 different projects ongoing in parallel) All this done by volunteers with actual paid jobs and life on the side ... so how much people/time/energy do you think remains to work on big-ass features that involves refactoring half of the project ?

So yeah, it's not exactly like we gonna read one comment, think "oh yeah that's right, we totally forgot about 2FA ?!?!", snap a finger and boom it's done ?

You could say "well, just tell me how much money it is worth and just hire somebody !" - but first of all, are you ready for the actual answer ? - and second, hiring somebody isn't gonna magically solve shit, because somebody has to train that person, review its work, handle the paperwork, etc. Personally I *hate* paperwork so not super enthusiast. But hey, if you'd like to handle paperwork and grants application etc, or if you have magic tips about recruiting long-term core contributors, we'll roll out the red carpet ! But like yeah we're caught in this big ass sprawling project which you could arguably be compared to Mozilla Firefox in terms of scope / importance for free internet (?) yet we're just a bunch of cyber-volunteering-hippies and who (most of us I think?) learned programming/sysadmin as a hobby ¯\\\_(ツ)\_/¯

But hey if that can give you some insight, I spent the day [trying to stabilize my PR on having several admins instead of one special admin user](https://github.com/YunoHost/yunohost/pull/1408), still not really stable yet and overall probably will take a full week of work (if including all the tests, user support etc that this will involve), and that's just one piece before ***maybe*** starting to think about the design (not even the implementation) of 2FA *for the admin* so ***maybe*** we could see it happening in 11.x (and yeah don't come ranting in 1 year being like "OMghGG u said 11.x but we're in 12.x now and still nothing amagad why so hard 1§!!11§??!!"). 2FA on the portal/SSO is another big ass story with several major refactorings along the way, certainly not happening before 12.x or 13.x.

**Comment by WoozyFace** [comment]
> @alexAubin Really ? Would love to know where exactly I said this .. (Who knows, maybe I did ...)



We'll not specifically but, it came down to a similar statement you did just now. I think we all know that there is no magic wand which solves things just with a spark of magic (except for women nowadays 😏). It does take your effort and time, and may I speak for all of us if I say that, you deserve your time to be alive.  
All I want and most of us is a secure workplace, why else is there a whole document on making things safer? But if you say that it has to do with a core element of Yunohost, then all we can do is help. Maybe we as (non-official)devs should team up somehow over discord or whatever and do some work for Alexandre instead of spamming all of his channels with questions. Let's fill it with answers. I cannot even imagine how much mail he must get from all the notifications. Sorry for this one. 

I hope that my idea would be appreciated.

**Comment by Mecallie** [comment]
> > how much work it is to enable 2FA (preferably via Oauth/SAML [...]) ? [...] why is it so hard to redesign the website?
> 
> I mean, I don't even know where to start with this ... "Why is it so hard to completely rework the entire SSO system of YunoHost and its 300 apps ecosystem ?" ... Not to mention that SSOs in general are one hell of a thing to engineer right.
> 
> I don't know if you folks realize how depressing this kind of comment is. The other day we had to deal _again_ with somebody _toxically_ complaining about how user self-registration mechanism is still not there and yet should have been the absolute top number one priority for 9 years (his words). And this was like the 5th time at least he was ranting about it.

My comment was not meant to depress anyone, it was an actual question: why is it so much work. And in that same post I even explained why I asked the question.

SSO and 2FA are two different things, although one might interfere with the other. The strength of Ynh lies in, I think, the fact that you can install apps in just one click and it all works together. Not in the SSO perse.

Does all the SSO integration need the https frontend that is used for the admin portal to work? To me it looks like it used the ldap db on localhost. Otherwise: why is it not an option to leave the current SSO in place, provide just the admin portal with 2FA based on something that can also be used for SSO (from within the apps, you do not need to redo all apps yourselves!) and gradually migrate to that?

That way you have about the same amount of work (bit less, since most apps can use an Idp for sso) but you can start more quickly and it will be pretty futureproof?

Again: I am not nagging or complaining (well, maybe a bit: sorry), but I would like to understand what makes it impossible to gradually start with this. The users/companies (non profits mostly) do not have any devs on the payroll, so I cannot provide any.

And again: thank you for the absolutely fantastic work on Yunohost. You are doing the community and the world in general a big favour! If I where a millionaire I would pay handsomely for all the work that has been done!

**Comment by zamentur** [comment]
## 2FA as high priority...
Personally, i think 2FA is not a high priority ticket in term of security. Considering security purpose, we have others todo more important currently and easier to achieve. 

And as explained by Aleks, 2FA required other PR to be finished first.

## Dangerous changes
I want to add that refactoring and improving stuff around authentication are a bit dangerous, for example Nextcloud had a big 2FA issues in the past due to an error in the implementation... And they have full time paid collaborators...

## Increase password requirement first
However, I agree that weak password on user account could be bruteforced and could mean a lot of spamming emails from your server AND a bad ip reputation. We should increase the requirements about password first of all, and that is more easy to do than 2FA.
For people who wants to increase yunohost security, it could be a good first issues: the code to adapt the strengh required is here: https://github.com/YunoHost/yunohost/blob/dev/src/utils/password.py#L42
And here
https://github.com/YunoHost/yunohost/blob/dev/src/settings.py#L66

## 2FA and conflicts with several use cases
Now each improvements on yunohost is more difficult due to the big number of uses cases. Students who were working on DNS stuff one year ago and in July 2022 gives the same feedback about it: it's difficult to list and keep in mind every use cases.

### Mixed internal and LDAP accounts
For example something not explain on this topic is several nextcloud mixed LDAP/SSO accounts with internal accounts. With SAML integration, how it works, could we upgrade or should we explain to this instance administrator that we deprecated their use cases ?

And Nextcloud is a well known app... What about others. It's already a mess between SSO/LDAP support...



### Multitenancy
Currently it's possible to configure an app as public and let user login with LDAP (or even SSO). It's convenient for multitenancy purpose, by this way user can login on a page with the same domain name than their app. Currently, YunoHost SSO is only on one domain, so it's not really convenient for all multitenancy purpose. 

SO if we force to login through a unique door, we have here a regression for some instances. The way to avoid this, could be to support several domain for SSO login page.

### The SSO has a "firewall" or a "guardian"
Some instance admins appreciate that SSOwat is between their app and the big Internet. In private mode for certains apps, it means that even if the app has a security hole (for example on a specific URL allowing to call a script in a library) the SSO will stop it, so the current SSOwat implementation could be seen as a kind of security measure against bot and app weakness...

It will be important i think to be able to keep this behavior in a potential new implementation. 

### Other uses cases already mentions...
... and use cases we miss.

So here if someone want to study the 350+ apps list (or just a part) to list which apps support what and how, let us know. We can provide you a collaborative onlyoffice sheet to document that.



## Is YunoHost currently made to protect very sensible data/life ?
Missing 2FA means for example you can be hacked being filmed by typing your password. Could be a jealous partner, a malicious relative or collaborator, a cops, an intelligence officer, a neonazi, a student or whatever...

It could be also a bot that grab your password on a leak password list or simply you use a too weak password indeed.

So yes, i agree with you 2FA could be useful in several use-cases. My feelings are several people wants 2FA by reflex but doesn't really need it if they have good numeric behavior, and others non tech users in specific situation should really have it. 

More generally, i think we miss  to explain people what yunohost can "guarantee" and what we cannot. And the answer is difficult cause it depends how you installed stuff too. But clearly, if you are afraid by intelligence agencies, cops or non trustable people around your server, we cannot guarantee a lot of things for now...

## How we move forward **and how you can help us**

- [ ] Finish the current PR about authentication and security
- [ ] Improve passwords strength
- [ ] Create 1 ticket for Webadmin and an other for SSO
- [ ] Document apps SAML support, if the app support mixed accounts
- [ ] **Documented** R&D on how implements 2FA inside yunohost without break all other features

**Comment by zamentur** [comment]
> Does all the SSO integration need the https frontend that is used for the admin portal to work? To me it looks like it used the ldap db on localhost. 

SSOwat currently doesn't depends of the Webadmin it's 2 well separated things. SSOwat use the password.py from YunoHost core API, but this should be change.

However, we have a PR started to make SSOwat running thanks to an a "user" API that could use the "admin" API, so in the future this 2 components could me coupled than today.

About apps here you have a list of apps use cases:
 - support only LDAP
 - support only SSO (via AUTH header)
 - support LDAP and SSO
 - + this 3 uses cases mixed with internal accounts (limesurvey, nextcloud and many others)


> Otherwise: why is it not an option to leave the current SSO in place, provide just the admin portal with 2FA based on something that can also be used for SSO (from within the apps, you do not need to redo all apps yourselves!) and gradually migrate to that?

I am not sure to understand the suggestion. But in all cases, we probably need indeed a migration scenario in which we could have app packaged in the old way and app package with the new way. And this will sadly take a looooong time i think. We already are in planning to support v2 and v3 app format during one or 2 years :/ at least.

**Comment by Mecallie** [comment]
> > Otherwise: why is it not an option to leave the current SSO in place, provide just the admin portal with 2FA based on something that can also be used for SSO (from within the apps, you do not need to redo all apps yourselves!) and gradually migrate to that?
> 
> I am not sure to understand the suggestion. But in all cases, we probably need indeed a migration scenario in which we could have app packaged in the old way and app package with the new way. And this will sadly take a looooong time i think. We already are in planning to support v2 and v3 app format during one or 2 years :/ at least.

What I mean by this is that:
1. Not all apps use SSO or can even use it.
2. A lot of apps already support built in SSO via an identity provider like KeyCloak << no need to rework those apps, just the script perhaps.
3. You can leave the existing SSO in place, but just put the admin portal behind 2FA.
4. If you do 3. you can use the built in 2FA from a lot of apps to provide more security. That is of no use if the admin/user portal does not have 2FA and you can just login with that using SSO.

As for the use cases for apps: Friendica, Nextcloud, Mastodon, Hubzilla and lots of others have built in support for SAML and 2FA. Connecting to the local ldap for that is not that big of an issue since it is not reachable from the internet directly.

And migration taking a lot of time: that's fine right? I mean: you can use the built-in security options from the app for 2FA if you want?

**Comment by alexAubin** [comment]
I mean we're here trying to demonstrate "why is this hard" and nobody can't even come up with a proper definition or scope of what we want to achieve ... This should be a self-sufficient demonstration of "why this is hard", not even mentioning that the current SSO is in Lua and nobody in the project likes to write Lua, let alone build a motherfucking SSO with 2FA in Lua, AND there's a gazillion other things to work on in parallel AND the number of regular core contributors can be counted on half of one hand.

4 apps supports SAML - great then ... We have no experience with SAML and the current SSO only works with Basic Auth. How de we design a general strategy than can be applied to all apps for which SSO / 2FA is relevant ? If we want to build something serious we should also investigate other SSO technologies and compare how they can be integrated, their pro and cons, etc.

I mean even for SAML I don't even know how that works but certainly there's more to it than just "connecting stuff to LDAP". What is the 2nd factor exactly ? Is it a OTP ? So one has to think about adding a new field somewhere in the LDAP DB to store it. Simple yeah ? Who is up to implement it ?

>And migration taking a lot of time: that's fine right?

Obviously you don't understand what you're talking about. Any developers that had to deal with migrating legacy stuff over X years will tell you that it is painful and boring as hell. And this tends to multiply the development time by a factor between 2 and 10.

***Even*** if this was a simple topic with no design challenge and critical security implications, the very fact that there's so much going on compared to the available volunteer time/energy means that this would take time.

I mean I don't even know what you're trying to achieve here. How many hours shall we continue to demonstrate that this is a complex topic and conclude that it won't happen anytime soon, rather than spending the same energy on actually working on stuff that will make it to the next releases ?

**Comment by zamentur** [comment]
If you want to document how install keycloak, and activate openid or whatever inside several apps to get 2FA, **feel free to do it**, we could probably publish this tutorial inside the doc. 

However, mixing openID, internal accounts, SSOwat/LDAP account could be quite confusing for final users :/ (ergonomy etc.). 
See for example Discourse or Limesurvey when 2 ways to login are actuvated, user never know on which part of the interface they should connect.

> I mean I don't even know what you're trying to achieve here. How many hours shall we continue to demonstrate that this is a complex topic and conclude that it won't happen anytime soon, rather than spending the same energy on actually working on stuff that will make it to the next releases ?

I agree with Aleks on that point. On my side, i consider i have given several todos to progress on prerequisites for 2FA on all yunohost instances, and has explain it's not my top priority currently (even my top priority on the security list of things to do). So at this step i hope everybody understand that this issues is difficult to solve, and that things could go faster if other persons try to progress on it by doing listed todos.

**Comment by renne** [comment]
I suggest to have a look at [py_webauthn](https://pypi.org/project/webauthn/) and [SimpleWebAuthn](https://github.com/MasterKale/SimpleWebAuthn).
By the way Passkey = Webauthn with private keys shared between devices.

**Comment by renne** [comment]
@alexAubin  @zamentur 
I want to try [WebAuthn proxy](https://github.com/Quiq/webauthn_proxy#nginx).
Will SSOwat play along with additional authentications in Nginx?

**Comment by Aeris1One** [comment]
I don't think it will, no :/

**Comment by sepp67** [comment]
Hello
My goal is to install MFA to Yunohost. As MFA solution, I am using Privacyidea. 

I found this plugin [LUA based NGINX Authentication plugin for privacyIDEA](https://github.com/dhoffend/lua-nginx-privacyidea).
it allows connecting Privacyidea to a NGINX server.
It is written in LUA. So all the conditions are here to work with the ssowat of Yunohost.

But I don't know how to implement it.
Which config file of NGINX in Yunohost should I Alter. The ssowat.conf or the global.conf or both ?

**Comment by renne** [comment]
@sepp67 
I asked about the SSO/user integration in the developer section of the Yunohost forum but never got an answer.
It seems there are no developers left who know how SSOwat works.
I'm looking for password-less Single-Sign-On with FIDO2 hardware security tokens. My dream would be to replace SSOwat with [Authentik](https://goauthentik.io/). That would not only allow to provide user data ([SCIM](https://scim.cloud/)) and SSO via Yunohost but also integrate Yunohost in existing authentication systems.
I also would love to see SSO in the Paperless-NGX app.
But no one seems tho know how SSOwat works and the maintainers don't want to replace SSOwat.
I'm seriously considering replacing Yunohost with some other solution as I do not see an up-to-date authentication system in Yunohost. SSOwat seems to be more tinkering around the edges than a secure solution.

**Comment by alexAubin** [comment]
@sepp67 / @renne 

**On the topic of "I found \<software X\> on the internet, how do I integrate it in nginx/ssowat"** :

:arrow_right: This is an issue tracker, not an advanced question forum. It's meant to discuss how to effectively integrate stuff in YunoHost as a project, not your personal instance. If your point is to integrate in YunoHost as a project, then there are many design considerations behind that kind of topic and this should be discussed with the team before jumping into implementation. Also we are in the process of refactoring ssowat which is going to take some time, and the priority probably is to first implement 2FA for the webadmin rather than 2FA for the user portal.

**On the topic of "no developers left who know how SSOwat works" and "the maintainers don't want to replace SSOwat"** : 

:arrow_right: Then you clearly don't know what you are talking about, or this is just a random aggressive statement to trigger volunteers of the project. This kind of toxic attitude won't be tolerated much longer.

See also [When will this feature be implemented? Why that app has not been packaged yet? I cannot believe you do not do that yet!](https://yunohost.org/en/community/faq#when-will-this-feature-be)

and also https://github.com/YunoHost/SSOwat/pull/217 and other related topics

**On the topic of "I'm considering replacing Yunohost" and "YunoHost doesn't seem like a secure solution"** :

:arrow_right: Then YunoHost is not secure and go ahead, use something else. We are volunteers and we do things in the hope that it may be useful for other people. We don't claim to be the best self-hosting solution there is. You are not a customer who pays for a product, and the project doesn't owe you anything. This issue tracker is not meant for you to rant about the project, and this kind of toxic attitude won't be tolerated much longer.

**Comment by csolisr** [comment]
Considering that the SSOwat refactor is now in place: what would be required to integrate a 2FA solution to the current stack?

**Comment by alexAubin** [comment]
Some way to create and store a OTP/TOTP related to an user, adapt the authentication mechanism of the yunohost-portal API and propagate this on the front ends

But honestly the priority is probably to implement it on the webadmin first

**Comment by zamentur** [comment]
We could use slapo-otp (not totally sure it fits our needs)
https://manpages.debian.org/unstable/slapd/slapo-otp.5.en.html

So i guess to achieve that, it means:

- [ ] **Check if with slapo-otp we could decide in which situation we want 2FA and in which not (password/info change, portal login, webadmin login)**
- [ ] Add the module slapo-otp and adapt slapd config
- [ ] Create a slapd migration
- [ ] Add a second (optional) step to the Webadmin login form
   - [ ] https://github.com/YunoHost/yunohost/blob/4e9d94c836fa34b37805bf07b61aa8aa0d9d61e3/src/authenticators/ldap_ynhuser.py#L196 
   - [ ] https://github.com/YunoHost/yunohost-admin/blob/4aba6a2a57f6fe3070fe2a7bab923d8d2874b728/app/src/views/LoginView.vue#L59
- [ ] Adapt UserCreate and UserEdit view
   - [ ] https://github.com/YunoHost/yunohost-admin/blob/4aba6a2a57f6fe3070fe2a7bab923d8d2874b728/app/src/views/user/UserCreate.vue#L79
   - [ ] https://github.com/YunoHost/yunohost/blob/4e9d94c836fa34b37805bf07b61aa8aa0d9d61e3/src/user.py#L266
   - [ ] https://github.com/YunoHost/yunohost-admin/blob/4aba6a2a57f6fe3070fe2a7bab923d8d2874b728/app/src/views/user/UserEdit.vue#L132
   - [ ] https://github.com/YunoHost/yunohost/blob/4e9d94c836fa34b37805bf07b61aa8aa0d9d61e3/src/user.py#L484

And it should be enough to get TOTP on webadmin, a bit more work to get it on UserPortal:
- [ ] Adapt Login to add a second step
   - [ ] https://github.com/YunoHost/yunohost/blob/4e9d94c836fa34b37805bf07b61aa8aa0d9d61e3/src/authenticators/ldap_ynhuser.py#L204
   - [ ] https://github.com/YunoHost/yunohost-portal/blob/034b312d7c14885fc783c7db843ccd57e846eaca/pages/login.vue#L33
- [ ] Adapt EditView
   - [ ] https://github.com/YunoHost/yunohost/blob/4e9d94c836fa34b37805bf07b61aa8aa0d9d61e3/src/portal.py#L238
   - [ ] https://github.com/YunoHost/yunohost-portal/blob/dev/components/UserPasswordForm.vue

**Some important questions:**
- [ ] How should work app ldap integration (a lot of apps may not include 2FA) with that ?
- [ ] What about apps using `basic-with-password` header ???

I wrote this in order to help people that have time or are motivate to implement it, i don't plan to implement it.

**Comment by eauchat** [comment]
Maybe this comment is a bit off topic (so feel free to mark it as so if you find it inappropriate), but I just wanted to specify the motivation for the like I added on @zamentur post.
I'm not so interested in 2FA being implemented in ynh, I see that could be useful for some people, and I think people for who it's a central thing have responsibility to engage in what they would like to see if that's not being done by other people who have other priorities.
I've followed this thread and the discussions in it. To me this post of @zamentur is one more of many examples I've seen in yunohost of how it's major contributors are engaging with requests from the community, and issues people have in a very proactive and committed way. Giving the opportunity to people to get involved and participate in improving things they care about in yunohost. I think it's pretty rare and precious and I appreciate the yunohost community and people who contribute to it for having this attitude, even when receiving undeserved criticism.
Spending the time to make such a detailed guide on how to implement something which is not necessarily a priority for you is very generous and showing lots of good will. I don't take it for granted and feel marveled many times a day when I interact with yunohost issues tracker, forum and the yunohost world and community.
Thanks to you and all core contributors, apps packagers, questions responders and more for that.

**Comment by eduardomozart** [comment]
> How should work app ldap integration (a lot of apps may not include 2FA) with that ?

I do believe it could work similarly as Sophos XG does (at least for OTP). The user informs password + 6 digits OTP into the *Password* field, then the YonoHost should strip the last 6 characters to check if LDAP user password is valid (if it binds) and, if it does, check the last 6 OTP digits is right. This way the LDAP only applications can still works as expected and no changes applied to them, as represented in the image below:

![Image](https://github.com/user-attachments/assets/d237afe6-e1ec-4b56-aeb9-f95aac872187)

But I do believe that YunoHost would need to implement a LDAP MFA Proxy or creates a LDAP server fork that does this as I couldn’t find any free LDAP server with this feature built in, except slapo-otp. It’s possible to differentiate users with OTP enabled through “oathOTPUser” attribute, if it exists is because user has OTP on LDAP set and it will work as in the image above (theoretically). More info at slapo-otp docs: https://man7.org/linux/man-pages/man5/slapo-otp.5.html

Here’s an howto: https://dev.to/a5hk/time-based-one-time-passwords-totp-with-openldap-5al0

But I’m not sure if slapo-otp approach would be desirable since it would enforce OTP even for 3rd party services like mail, which would sucks typing the password with OTP into the mail client every time since the password changes every 30 seconds. I couldn’t find any slapo overlay or slapd-sock binding socket that would allow to create exception for specific clients like PAM modules allows to. The slapd also have the slapo-nssov module which also could be used to implement OTP by relying on pam_oath module, but it would face the same issue of slapo-otp since there’s seems to be no way to disable (not enforce) OTP for specific LDAP applications.

Maybe a workaround would be creating two LDAP databases, one with slapo-otp and another without, and sync them, so the default LDAP server (:631) requires OTP (when user has it setup) while another LDAP server listening on another port does not, so we could use it for apps which OTP maybe not desirable (like mail service). It could use https://serverfault.com/questions/559517/issues-with-ldap-backend-in-openldap slapd module/plugin to sync the LDAP database and maybe back_meta slapd module to redirect LDAP requests for each LDAP server. The use of multiple OpenLDAP instances into a single server is documented at https://www.vincentliefooghe.net/content/openldap-setup-multiple-instances-server and this approach would allow add an option to YunoHost app catalog installer to enforce OTP or not (this switch would simply change the LDAP port for the app). A LDAP MFA Proxy would not require a second LDAP database and sync, but I was unable to find any free/open source software for this, only DUO LDAP Proxy (free for up 10 users).

Other way to make it work would be using LemonLDAP (LLNG) for Web Apps instead of relying on SSOWat - it seems a mature enough solution for YunoHost instead of maintaining and refactor SSOWat itself.

There’s DUO LDAP Proxy that sends a notification to authorize the login to the user smartphone (no OTP code needed), but it’s free for only 10 users and for YunoHost provide similar feature it would requires paying for Apple to send notifications for devices.
